### ISSUE TITLE
Determine why there appears to be a shift in the position of sources on the focal plane.

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#14]
As discussed in PR #4.

I think I will try to make a single source at the center of the focal plane and see what happens.

**Comment by cwwalter** [comment]
I studied this by placing a single star at the same position as the telescope pointing and checking whether it was imaged at the center of the sensor with ds9.  I also looked at the fits headers.

There are at least two effects happening which cause this:

- PhoSim:  The Twinkles instance files include the default perturbations to the optics etc.  This can vary positions by the order of 10 pixels.  There was in fact a 6-7 pixel shift in X and Y.  When I cleared all perturbations this went away.  Thanks to @SimonKrughoff for helping with this! There is actually still a +1 +1 shift but I think the actual center maybe at a pixel intersection and I think is is OK.

For reference the FITS header info looks like:

```
CRPIX1  =    2033.776902985953                                                  
CRPIX2  =    2000.960606544964                                                  
CRVAL1  =           31.1133844                                                  
CRVAL2  =           -10.097006      
```

- imSim: There is a +1 +3 pixel shift.  However if you look at the FITS headers it contains this:

```
CRPIX1  =    2033.776902985953                                                  
CRPIX2  =    2000.960606544964                                                  
CRVAL1  =           31.1133844                                                  
CRVAL2  =           -10.097006   
```

In other words, the star is in fact where the system thinks the center is.

I am not sure it is relevant but this GalSim WCS related code (which I don't yet understand) seems to be doing a transformation which modifies the position:

```python
def _newOrigin(self, origin):
        """
        This is a method required by the GalSim WCS API.  It returns
        a copy of self, but with the pixel-space origin translated to a new
        position.
        @param [in] origin is an instantiation of a galsim.PositionD representing
        the a point in pixel space to which you want to move the origin of the WCS
        @param [out] _newWcs is a WCS identical to self, but with the origin
        in pixel space moved to the specified origin
        """
        _newWcs = GalSim_afw_TanSipWCS(self.afwDetector, self.afwCamera, self.obs_metadata, self.epoch,
                                       photParams=self.photParams, wcs=self._tanSipWcs)
        _newWcs.crpix1 = origin.x
        _newWcs.crpix2 = origin.y
        _newWcs.fitsHeader.set('CRPIX1', origin.x)
        _newWcs.fitsHeader.set('CRPIX2', origin.y)
        return _newWcs
```

@danielsf or @rmjarvis can you comment?

**Comment by cwwalter** [comment]
One consequence of this is that we will always see a shift between PhoSim and imSim with the current Twinkles instance files.  I think it will likely also be focal plane position dependent.

**Comment by danielsf** [comment]
I don't really know anything more than the docstring says.  GalSim expects its WCSs to have a method like this.  I am not sure how they are used.

**Comment by rmjarvis** [comment]
Comparing this:
```
def _newOrigin(self, origin):
        """
        This is a method required by the GalSim WCS API.  It returns
        a copy of self, but with the pixel-space origin translated to a new
        position.
        @param [in] origin is an instantiation of a galsim.PositionD representing
        the a point in pixel space to which you want to move the origin of the WCS
        @param [out] _newWcs is a WCS identical to self, but with the origin
        in pixel space moved to the specified origin
        """
        _newWcs = GalSim_afw_TanSipWCS(self.afwDetector, self.afwCamera, self.obs_metadata, self.epoch,
                                       photParams=self.photParams, wcs=self._tanSipWcs)
        _newWcs.crpix1 = origin.x
        _newWcs.crpix2 = origin.y
        _newWcs.fitsHeader.set('CRPIX1', origin.x)
        _newWcs.fitsHeader.set('CRPIX2', origin.y)
```
to the implementation of `_newOrigin` in the `GSFitsWCS` class:
```
def _newOrigin(self, origin):
        ret = self.copy()
        if origin is not None:
            ret.crpix = ret.crpix + [ origin.x, origin.y ]
        return ret
```
I think the problem may be that it is ignoring the existing value of crpix1,2.  I think the correct implementation should be:
```
        if origin is not None:
            _newWcs.crpix1 += origin.x
            _newWcs.crpix2 += origin.y
            _newWcs.fitsHeader.set('CRPIX1', _newWcs.crpix1)
            _newWcs.fitsHeader.set('CRPIX2', _newWcs.crpix2)
```
However, I'm not at all sure that this the proximate cause of the shift you are investigating.

**Comment by rmjarvis** [comment]
Previous comment updated on Github...

**Comment by cwwalter** [comment]
So just a bit of information... Making Mike's suggested change does not actually result in changing either the values in the fits files or where the star shows up on the sensor.  However, printing this:

`(self.crpix1, self.crpix2),  (origin.x, origin.y),  (self.crpix1-origin.x, self.crpix2-origin.y)`

[Note the "-" not "+"]

gives

`(2033.77, 2000.96) (-3.22, -0.039) (2037.0, 2001.0)`

So the origin is in fact shifted by an amount that moves you off the center (which it turns out is also wrong by one pixel in each dimension).  I will follow up more...

**Comment by cwwalter** [comment]
@danielsf I'm closer to figuring out what is going on here but have a question.

In 

in galSimInterpreter.py when we get here:

```python 
for detector in detectorList:

                name = self._getFileName(detector=detector, bandpassName=bandpassName)

                xPix, yPix = pixelCoordsFromPupilCoords(numpy.array([gsObject.xPupilRadians]),
                                                        numpy.array([gsObject.yPupilRadians]),
                                                        chipNames=[detector.name],
                                                        camera=detector.afwCamera)
```

I find for a centered star that the gsObject.xPupilRadians and gsObject.yPupilRadians are (0,0) which I assume means centered down the bore.  But, the returned (xPix, yPix) are not at the center of the chip:

```
ipdb> print (xPix, yPix)
(array([ 2032.789]), array([ 1999.952]))
```

I was going to start debugging with the code in:

https://github.com/lsst/sims_coordUtils/blob/master/python/lsst/sims/coordUtils/CameraUtils.py

But I see you have made tons of commits lately.  Does this sound like a bug you might have fixed?

**Comment by danielsf** [comment]
No, I have not addressed that bug.  The conversion from pupil coordinates to pixel coordinates is handled by a lot of `afw` code.  I'm not exactly sure how that his tested and validated.  If there was something wrong in there, I'm not sure we (where "we" means "sims") would know.

**Comment by danielsf** [comment]
It might be worth asking the people in charge of `obs_lsstSim`, which actually contains the software representation of our camera, if this is expected behavior or not.  I'm not actually sure who is in charge of `obs_lsstSim`, so it's probably deserving of a community post.

**Comment by cwwalter** [comment]
> It might be worth asking the people in charge of obs_lsstSim, which actually contains the software representation of our camera, if this is expected behavior or not. I'm not actually sure who is in charge of obs_lsstSim, so it's probably deserving of a community post.

I think it is @SimonKrughoff.  BTW, do you agree with my interpretation of what should be happening? I mean that ra, dec = (0,0) should be resulting in the center of the sensor?

**Comment by danielsf** [comment]
@cwwalter One quick thought: what happens when you set the `includeDistortion` kwarg to `pixelCoordsFromPupilCoords` equal to `False`?

**Comment by cwwalter** [comment]
Good idea... but it didn't change anything.

**Comment by cwwalter** [comment]
Following up on a suggestion by @SimonKrughoff, it turns out these offsets were due to the chip offsets and rotations that are embedded in the model described in the file focalplanelayout.txt supplied by PhoSim.  PhoSim uses this file itself, and it is used by imSim and the LSST stack via the makeLsstCameraRepository.py script in the obs_lsstSim module which builds the camera model.

On running makeLsstCameraRepository.py (which is done when building the package with scons) the program parses focalplanelayout.txt and builds the camera model which is encoded in the python code output in description/camera/camera.py.

The relevant columns in focalplanelayout.txt are:

 ```
2:  x position in microns
3:  y position in microns
10: 1st Euler Rotation in degrees
13: x translation in mm
14: y translation in mm
```

The makeLsstCameraRepository.py writes the geometry file using

```python
detConfig.offset_x = float(els[1])/1000. + float(els[12])
detConfig.offset_y = float(els[2])/1000. + float(els[13])
```

and

```python
detConfig.yawDeg = 90.*nQuarter + float(els[9])
```

Thus encoding the final net position in the model.  In PhoSim use of the clearpertubation or perturbation mode commands can give you the ideal model with no shifts.  Currently this is not possible with imSim due to the way obs_lsstSim was designed.  

I confirmed this solution by modifying makeLsstCameraRepository.py to not apply the offsets and the shifts went away.

In conclusion: this effect is understood.  For imSim, position comparisons must be done after astrometric corrections unless a modified version of obs_lsstSim is used.  Similar considerations must be given to compare against PhoSim.  In principle if both PhoSim and imSim are using the same model the offsets should agree. However other optical perturbations in the PhoSim model might cause differences.

I am closing this issue.  A possible new issue would be allow for simulation an "ideal" focalplane in imSim.
